lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,248 @@
+	if err != nil {
+		return trace.BadParameter("Redis connection string is incorrect: %v", err)
+	}
+
+	var (
+		redisConn      redis.UniversalClient
+		connectionAddr = fmt.Sprintf("%s:%s", connectionOptions.address, connectionOptions.port)
+	)
+
+	// TODO(jakub): Use system CA bundle if connecting to AWS.
+	// TODO(jakub): Investigate Redis Sentinel.
+	switch connectionOptions.mode {
+	case Standalone:
+		redisConn = redis.NewClient(&redis.Options{
+			Addr:      connectionAddr,
+			TLSConfig: tlsConfig,
+		})
+	case Cluster:
+		if sessionCtx.DatabaseName != "" {
+			// ref: https://redis.io/commands/select
+			return trace.BadParameter("Redis Cluster only supports database zero")
+		}
+
+		redisConn = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs:     []string{connectionAddr},
+			TLSConfig: tlsConfig,
+		})
+	default:
+		// We've checked that while validating the config, but checking again can help with regression.
+		return trace.BadParameter("incorrect connection mode %s", connectionOptions.mode)
+	}
+
+	defer func() {
+		if err := redisConn.Close(); err != nil {
+			e.Log.Errorf("failed to close Redis connection: %v", err)
+		}
+	}()
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	if err := e.process(ctx, redisConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// readClientCmd reads commands from connected Redis client.
+func (e *Engine) readClientCmd(ctx context.Context) (*redis.Cmd, error) {
+	cmd := &redis.Cmd{}
+	if err := cmd.ReadReply(e.clientReader); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	val, ok := cmd.Val().([]interface{})
+	if !ok {
+		return nil, trace.BadParameter("failed to cast Redis value to a slice")
+	}
+
+	return redis.NewCmd(ctx, val...), nil
+}
+
+// process is the main processing function for Redis. It reads commands passed from client and passes them to
+// a Redis instance. It's also responsible for audit.
+func (e *Engine) process(ctx context.Context, redisClient redis.UniversalClient) error {
+	for {
+		// Read commands from client.
+		cmd, err := e.readClientCmd(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// send valid commands to Redis instance/cluster.
+		err = e.processCmd(ctx, redisClient, cmd)
+
+		var vals interface{}
+		if _, ok := err.(redis.Error); ok {
+			vals = err
+		} else if errors.Is(err, context.DeadlineExceeded) {
+			// Do not return Deadline Exceeded to the client as it's not very self-explanatory.
+			// Return "connection timeout" as this is what most likely happened.
+			vals = trace.ConnectionProblem(err, "connection timeout")
+		} else if err != nil {
+			// Terminate connection only on connection errors. Errors that comes from Redis
+			// should be propagated back to the client.
+			if utils.IsOKNetworkError(err) {
+				return trace.Wrap(err)
+			}
+
+			vals = err
+		} else {
+			vals, err = cmd.Result()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
```

Comment:
Can this block be refactored a little so it follows a more "traditional" error handling flow, or even moved into a separate method?

```go
if err != nil {
  return e.processClientError(err)
}
vals, err := cmd.Result()
...
return e.sendToClient(vals)
```

Reply:
I've refactored this part. Let me know what do you think.